### PR TITLE
Added GaussianProcessesStrategy

### PR DESCRIPTION
**Pull request by joellidin**
This pull request adds the following:
* A new strategy using GP-UCB
* Added a small example for this strategy
* Made `on-update` callback a bit more robust in `thompsons_sampling.py`

### CODE REVIEW

File: thompson_sampling/thompson_strategy.py [hunk 1]
```diff
@@ -78,3 +89,51 @@ def mean(self):
     @property
     def selections(self):
         return self.num_selections
+
+
+class GaussianProcessesStrategy(ThompsonStrategy):
+    def __init__(
+        self,
+        num_parameters: int,
+        interval: "np.ndarray | list[np.ndarray]",
```
Comment (rssalessio) [inline thompson_sampling/thompson_strategy.py]: shouldn't this be Union[np.ndarray, list[np.ndarray]]?

File: thompson_sampling/thompson_strategy.py [hunk 2]
```diff
@@ -78,3 +89,51 @@ def mean(self):
...
+        if isinstance(interval, list):
+            if len(interval) != self.num_parameters:
+                raise ValueError(
+                    "When passing a list the length should be the same as num_parameters"
+                )
+            else:
+                self.mesh = np.meshgrid(*interval)
+
+        elif isinstance(interval, np.ndarray):
+            self.mesh = np.meshgrid(*[interval for _ in range(self.num_parameters)])
```
Comment (rssalessio) [inline thompson_sampling/thompson_strategy.py]: the interval could be different for each parameter.

File: thompson_sampling/thompson_strategy.py [hunk 3]
```diff
@@ -78,3 +89,51 @@ def mean(self):
...
+
+        mapfunc = partial(np.ravel, order="A")
```
Comment (rssalessio) [inline thompson_sampling/thompson_strategy.py]: why not a lambda function?

File: thompson_sampling/thompson_strategy.py [hunk 4]
```diff
@@ -78,3 +89,51 @@ def mean(self):
...
+        self.X_grid = np.column_stack(list(map(mapfunc, self.mesh)))
+
+        self.mu = np.array([0.0 for _ in range(self.X_grid.shape[0])])
```
Comment (rssalessio) [inline thompson_sampling/thompson_strategy.py]: [0.] * len(self.X_grid)

File: thompson_sampling/thompson_strategy.py [hunk 5]
```diff
@@ -78,3 +89,51 @@ def mean(self):
...
+        self.sigma = np.array([np.random.randn() for _ in range(self.X_grid.shape[0])])
```
Comment (rssalessio) [inline thompson_sampling/thompson_strategy.py]: i believe it's better to keep this fixed at some value..maybe it should be a parameter of init

File: thompson_sampling/thompson_strategy.py [hunk 6]
```diff
@@ -78,3 +89,51 @@ def mean(self):
...
+        self.beta = beta
+        kernel = C(1, (1e-3, 1e1)) * RBF(1, (1e-2, 1e1)) + WhiteKernel()
```
Comment (rssalessio) [inline thompson_sampling/thompson_strategy.py]: I believe a better option is to pass the kernel to this class. in this way you don't have to define it here. Maybe, even better, a user can define its own gaussian process regressor and pass it to this function.

File: thompson_sampling/thompson_strategy.py [hunk 7]
```diff
@@ -78,3 +89,51 @@ def mean(self):
...
+        self.gp = GaussianProcessRegressor(kernel=kernel)
+        self.X = []
+        self.y = []
+
+    def sample(self):
+        # if len(self.X) < 25:
+        #     grid_idx = np.random.randint(self.X_grid.shape[0])
+        # else:
+        grid_idx = (self.mu.flatten() + self.sigma * np.sqrt(self.beta)).argmax()
```
Comment (rssalessio) [inline thompson_sampling/thompson_strategy.py]: beta needs to be a callable, maybe function of x?

File: thompson_sampling/thompson_strategy.py [hunk 8]
```diff
@@ -78,3 +89,51 @@ def mean(self):
...
+        params = self.X_grid[grid_idx]
+        return params
+
+    def update(self, reward, params):
+        self.X.append(params)
+        self.y.append(reward)
+        if len(self.X) < 25:
+            return
+        self.gp = self.gp.fit(self.X, self.y)
+        self.mu, self.sigma = self.gp.predict(self.X_grid, return_std=True)
```
Comment (rssalessio) [inline thompson_sampling/thompson_strategy.py:158]: if here you compute mu, maybe you should also compute beta as funtion of X_grid

File: thompson_sampling/thompson_strategy.py [hunk 9]
```diff
@@ -78,3 +89,51 @@ def mean(self):
...
+        #     grid_idx = np.random.randint(self.X_grid.shape[0])
+        # else:
+        grid_idx = (self.mu.flatten() + self.sigma * np.sqrt(self.beta)).argmax()
+        params = self.X_grid[grid_idx]
+        return params
+
+    def update(self, reward, params):
+        self.X.append(params)
+        self.y.append(reward)
+        if len(self.X) < 25:
```
Comment (rssalessio) [inline thompson_sampling/thompson_strategy.py]: why 25?

File: thompson_sampling/thompson_sampling.py [hunk 10]
```diff
@@ -38,29 +42,31 @@ def __init__(self,
...
-                    if 'on_update' in self.callbacks:
-                        if self.callbacks['on_update'](
-                                epoch, self.strategy,
-                                epoch == self.epochs - 1):
+                    if "on_update" in self.callbacks:
+                        epoch_tmp = np.round(epoch / self.num_processors, 5)
+                        if self.callbacks["on_update"](
+                            epoch_tmp,
+                            self.strategy,
+                            (self.epochs - 1 / self.num_processors) - epoch_tmp < 1e-5,
```
Comment (rssalessio) [inline thompson_sampling/thompson_sampling.py]: use np.isclose(self.epochs - 1 / self.num_processors, epoch_tmp)

File: thompson_sampling/thompson_strategy.py (hunk 2 above)
Comment (joellidin) [reply thompson_sampling/thompson_strategy.py]: You can pass the interval either as a single `np.ndarray`, giving the same interval for all the parameters. Or, you can pass a list of `np.ndarray` of length self.num_parameters, which could all be different.

File: thompson_sampling/thompson_strategy.py (hunk 5 above)
Comment (joellidin) [reply thompson_sampling/thompson_strategy.py]: fixed in 94ba55000419022918e29b985e505e5bc109131b

File: thompson_sampling/thompson_strategy.py (hunk 4 above)
Comment (joellidin) [reply thompson_sampling/thompson_strategy.py]: fixed in 94ba55000419022918e29b985e505e5bc109131b

File: thompson_sampling/thompson_strategy.py (hunk 2 above)
Comment (rssalessio) [reply thompson_sampling/thompson_strategy.py]: ok, something like *([interval] * num_parameters)

File: thompson_sampling/thompson_strategy.py (hunk 9 above)
Comment (joellidin) [reply thompson_sampling/thompson_strategy.py]: fixed in 94ba55000419022918e29b985e505e5bc109131b

File: thompson_sampling/thompson_strategy.py (hunk 6 above)
Comment (joellidin) [reply thompson_sampling/thompson_strategy.py]: fixed in 94ba55000419022918e29b985e505e5bc109131b

File: thompson_sampling/thompson_strategy.py (hunk 3 above)
Comment (joellidin) [reply thompson_sampling/thompson_strategy.py]: removed mapfunc in 94ba55000419022918e29b985e505e5bc109131b

File: thompson_sampling/thompson_strategy.py (hunk 2 above)
Comment (joellidin) [reply thompson_sampling/thompson_strategy.py]: fixed in 94ba55000419022918e29b985e505e5bc109131b

File: thompson_sampling/thompson_strategy.py (hunk 1 above)
Comment (joellidin) [reply thompson_sampling/thompson_strategy.py]: fixed in 94ba55000419022918e29b985e505e5bc109131b

File: thompson_sampling/thompson_sampling.py (hunk 10 above)
Comment (joellidin) [reply thompson_sampling/thompson_sampling.py]: fixed in ea43096ed4d8dd42aad9fd3ef9bf3d96a216b6ec

**Pull request merged**